File: leetcode/5465.number-of-nodes-in-the-sub-tree-with-the-same-label.py
import collections


class Solution:
    def countSubTrees(self, n: int, edges: List[List[int]], labels: str) -> List[int]:
        tree = collections.defaultdict(list)
        for f, t in edges:
            tree[f].append(t)
            tree[t].append(f)

        ans = [0] * n

        def dfs(node, parent):
            d = collections.Counter()
            d[labels[node]] = 1
            for child in tree[node]:
                if child == parent:
                    continue
                dc = dfs(child, node)
                for la in dc:
                    d[la] += dc[la]
            ans[node] = d[labels[node]]
            return d

        dfs(0, None)
        return ans

File: leetcode/5465.number-of-nodes-in-the-sub-tree-with-the-same-label.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {string} labels
 * @return {number[]}
 */
var countSubTrees = function (n, edges, labels) {
  // TLE
  const tree = Array(n)
    .fill(0)
    .map(() => [])

  edges.forEach(([f, t]) => {
    tree[f].push(t)
    tree[t].push(f)
  })

  const ans = Array(n)
  const dfs = (i, p) => {
    const l = { [labels[i]]: 1 }
    for (let j = 0; j < tree[i].length; j++) {
      const child = tree[i][j]
      if (child === p) {
        continue
      }
      const lc = dfs(child, i)
      Object.keys(lc).forEach((k) => (l[k] = (l[k] || 0) + lc[k]))
    }
    ans[i] = l[labels[i]]
    return l
  }

  dfs(0)
  return ans
}
